File: step2.R
library(httr)
library(digest)
library(XML)
library(base64enc)

uid <- Sys.getenv("LABARCHIVES_UID")
notebook_id <- Sys.getenv("LABARCHIVES_NOTEBOOK_ID")
key_id <- Sys.getenv("LABARCHIVES_KEY_ID")
access_password <- Sys.getenv("LABARCHIVES_ACCESS_PASSWORD")

required <- c(
  LABARCHIVES_UID = uid,
  LABARCHIVES_NOTEBOOK_ID = notebook_id,
  LABARCHIVES_KEY_ID = key_id,
  LABARCHIVES_ACCESS_PASSWORD = access_password
)
if (any(required == "")) {
  stop("Missing required LabArchives environment variables.")
}

time_call <- paste0("https://api.labarchives.com/api/utilities/epoch_time?akid=", key_id)
time_response <- GET(time_call)
stop_for_status(time_response)
expires <- xmlValue(xmlRoot(xmlParse(content(time_response, "text")))[[1]])

sig_raw <- paste0(key_id, "notebook_backup", expires)
sig_digest <- hmac(charToRaw(access_password), charToRaw(sig_raw), algo = "sha1", raw = TRUE)
sig <- URLencode(base64encode(sig_digest), reserved = TRUE)

response <- GET(
  "https://api.labarchives.com/api/notebooks/notebook_backup",
  query = list(
    uid = uid,
    nbid = notebook_id,
    akid = key_id,
    expires = expires,
    sig = sig
  )
)
stop_for_status(response)

writeBin(content(response, "raw"), "notebook_backup_response.bin")

File: step2.py
import base64
import hmac
import os
import urllib.parse
from hashlib import sha1
from xml.etree import ElementTree as ET

import requests

UID = os.environ.get("LABARCHIVES_UID")
NOTEBOOK_ID = os.environ.get("LABARCHIVES_NOTEBOOK_ID")
KEY_ID = os.environ.get("LABARCHIVES_KEY_ID")
ACCESS_PASSWORD = os.environ.get("LABARCHIVES_ACCESS_PASSWORD")

required = {
    "LABARCHIVES_UID": UID,
    "LABARCHIVES_NOTEBOOK_ID": NOTEBOOK_ID,
    "LABARCHIVES_KEY_ID": KEY_ID,
    "LABARCHIVES_ACCESS_PASSWORD": ACCESS_PASSWORD,
}
missing = [name for name, value in required.items() if not value]
if missing:
    raise RuntimeError(f"Missing required environment variables: {', '.join(missing)}")

time_response = requests.get(
    f"https://api.labarchives.com/api/utilities/epoch_time?akid={KEY_ID}",
    timeout=30,
)
time_response.raise_for_status()
expires = ET.fromstring(time_response.text)[0].text

sig_raw = f"{KEY_ID}notebook_backup{expires}".encode()
sig_digest = hmac.new(ACCESS_PASSWORD.encode(), sig_raw, digestmod=sha1).digest()
sig = urllib.parse.quote(base64.b64encode(sig_digest), safe="")

params = {
    "uid": UID,
    "nbid": NOTEBOOK_ID,
    "akid": KEY_ID,
    "expires": expires,
    "sig": sig,
}

response = requests.get(
    "https://api.labarchives.com/api/notebooks/notebook_backup",
    params=params,
    timeout=120,
)
response.raise_for_status()

with open("notebook_backup_response.bin", "wb") as output:
    output.write(response.content)
